### PR TITLE
Avoid delete all the related content

### LINKED ISSUE #20491: Many-to-many relationships between a single content type removes relationships on save/publish
As reported in https://dotcms.zendesk.com/agent/tickets/104279 v5.3.8.4
Recreated in demo v21.05

If you have a content type that has a many-to-many relationship to itself it can cause other contentlets to lose that relationship. 

The log does not throw any errors.

1. Create a content type with a many-to-many relationship to itself.
2. Create 3 contentlets, A, B, C
3. On A, add a relationship to B and C - save/publish
4. On B add a relationship to A and C - save/publish
5. On C add a relationship to A and B - save/publish
6. Reopen A, the relationship to C has been lost
7. Re-add C to A - save/publish
8. Reopen B, all relationships are lost.
9. Re-add A and C to B - save/publish... The relationships appear to hold at this point. But you get an unsaved changes warning when you re-open even if you don't make changes.
10. However if you add a relationship to a different content - D - the relationships break down again.

STEP 3
image

STEP 4
image

STEP 5
image

STEP 6
image

STEP 7
image

STEP 8
image

STEP 9
image

RELATIONSHIPS HOLDING
image

image

image

BUT YOU GET THE UNSAVED MESSAGE WARNING
image

STEP 10
image

RELATIONSHIPS ARE LOST AGAIN
image

### PR DESCRIPTION
**Pull request by freddyucv**
Avoid delete all the related conten

### CODE REVIEW

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 1]
```diff
@@ -746,4 +748,80 @@ private static AcknowledgedResponse setClusterAsReadOnly(final boolean value) {
...
+
+        assertEquals(2, contentlets.size());
+
+        for (Contentlet contentletRelated : contentsRelated) {
+            assertTrue(contentlets.contains(contentletRelated.getIdentifier()));
+        }
+    }
+
+    @NotNull
+    private Contentlet releteContent(
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:818]: rename this method please, it should be `relateContent`

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 2]
```diff
@@ -746,4 +748,80 @@ private static AcknowledgedResponse setClusterAsReadOnly(final boolean value) {
...
+        Contentlet contentletC = new ContentletDataGen(contentType)
+                .host(host)
+                .nextPersisted();
+
+        contentletA = releteContent(relationship, contentletA, contentletB, contentletC);
+        contentletB = releteContent(relationship, contentletB, contentletA, contentletC);
+        contentletC = releteContent(relationship, contentletC, contentletA, contentletB);
+
+        contentletA.setProperty(relationship.getChildRelationName(), Arrays.asList(contentletB, contentletC));
+        contentletB.setProperty(relationship.getChildRelationName(), Arrays.asList(contentletA, contentletC));
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:791]: why a re you setting the relationship for contentletB and C if only A is checked in?

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 3]
```diff
@@ -3600,8 +3601,17 @@ public void deleteRelatedContent(final Contentlet contentlet, final Relationship
                 .dbRelatedContent(relationship, contentlet, hasParent);
         cons = permissionAPI
                 .filterCollection(cons, PermissionAPI.PERMISSION_READ, respectFrontendRoles, user);
-        APILocator.getRelationshipAPI().deleteByContent(contentlet, relationship, cons);
 
+        for (final Contentlet relatedContent : cons) {
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java:3605]: Does this logic consider non-self related content?

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java (hunk 1 above)
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java]: done https://github.com/dotCMS/core/pull/20558/commits/fa048b2c56dcf34de41da210dfa35a2e2c0224b5#diff-093de07a53a725b587963fd60c024f52e089c4f78e7ea88ba8e3a3a0cdaecab9R816

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java (hunk 2 above)
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java]: done https://github.com/dotCMS/core/pull/20558/commits/fa048b2c56dcf34de41da210dfa35a2e2c0224b5#diff-093de07a53a725b587963fd60c024f52e089c4f78e7ea88ba8e3a3a0cdaecab9L791

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java (hunk 3 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java:3605]: The bug not happen for non-self related

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 4]
```diff
@@ -746,4 +749,77 @@ private static AcknowledgedResponse setClusterAsReadOnly(final boolean value) {
...
+    }
+
+    private Contentlet relateContent(
+            Relationship relationship,
+            Contentlet parentContent,
+            Contentlet... contentletChilds) {
+        final Contentlet parentCheckout = ContentletDataGen.checkout(parentContent);
+        parentCheckout.setProperty(relationship.getChildRelationName(), Arrays.asList(contentletChilds));
+        final Contentlet checkin =  ContentletDataGen.checkin(parentCheckout);
+        return ContentletDataGen.checkout(checkin);
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:823]: After discussion this `checkout` should be moved outside of this method

Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java]: done https://github.com/dotCMS/core/pull/20558/commits/ba3d5e566ac3b7b4dc6c03b815a0aa8a748fa7b7

**Review by dsilvam — APPROVED** [review]

**Review by nollymar — APPROVED** [review]

**Pull request merged**